crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4631,6 +4633,25 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_invalid_type_expression(
+        &mut self,
+        expression: &ast::Expr,
+        error: InvalidTypeExpressionError<'db>,
+    ) -> Type<'db> {
+        let InvalidTypeExpressionError {
+            fallback_type,
+            invalid_expressions,
+        } = error;
+        for error in invalid_expressions {
+            self.diagnostics.add_lint(
+                &INVALID_TYPE_FORM,
+                expression.into(),
+                format_args!("{}", error.reason()),
+            );
+        }
+        fallback_type
+    }
```

Comment:
I feel like I slightly prefer the pattern we use in `CallOutcome` etc, where we have a method on the result type (or could just be on the error variant) that takes in a DiagnosticsBuilder and a node and emits the right diagnostics, returning the fallback type. This makes handling the result a bit more ergonomic (though you do have to explicitly pass in the diagnostics builder), and it prevents proliferating `infer_*` methods in `TypeInferenceBuilder` (which is already too large) that don't really fit the pattern of "infer a type for an AST node."

I realize this means passing AST nodes into `types.rs`, but I see the result/outcome types as kind of an intermediate layer between type inference builder and `Type` operations, that brings together nodes, diagnostics, and type operation results. I already moved `CallOutcome` out of `types.rs`; we could move more of those result/outcome types out of `types.rs` to clarify this layering.

This is all kind of musing, though; open to arguments for the pattern you used here.

Reply:
I do feel like a `Result` works really well here. There really are only two possibilities -- either it's okay, and we don't need to emit a diagnostic, or it isn't, and we do. If I were to write a custom `*Outcome`-like enum here, I think it would end up looking exactly like `Result`, so reinventing the wheel feels a little silly. And I just don't know what I'd call the enum 😆

But I totally take the point that the logic for unwrapping the error struct into a type (and emitting a diagnostic in the process) doesn't need to be on the `TypeInferenceBuilder`. I've moved it into an `InvalidTypeExpressionError::into_fallback_type()` method instead. I agree it's much nicer that way, and means we'll be able to easily move the enum into another submodule in the future if we want to.